### PR TITLE
Don't pull out the full state when calculating push actions

### PR DESCRIPTION
**Pull request by erikjohnston**
Fixes https://github.com/matrix-org/synapse/issues/12653

### CODE REVIEW

File: synapse/push/bulk_push_rule_evaluator.py [hunk 1]
```diff
@@ -298,10 +302,19 @@ async def action_for_event_by_user(
...
+            # event from before the user joined, which they should not be pushed
+            # for
+            visible = await filter_events_for_client_with_state(
+                self.store, uid, event, context
+            )
+            if not visible:
+                continue
+
+            localpart = get_localpart_from_id(uid)
+            profile_info = await self.store.get_profileinfo(localpart)
```
Comment (erikjohnston) [inline synapse/push/bulk_push_rule_evaluator.py:316]: This gets the display name from the profile table (i.e. the users global profile), rather than the per room profile. It's also not cached.

I don't think we have an existing DB function for getting per room profiles, but I think we can join `current_state_events` and `room_memberships` to get the display name out.

We should also make this a batched look up before the loop.

File: synapse/push/bulk_push_rule_evaluator.py [hunk 2]
```diff
@@ -452,145 +465,34 @@ async def get_rules(
...
-                push_rules_delta_state_cache_metric.inc_misses()
-            # Ensure the state IDs exist.
-            assert current_state_ids is not None
-
-            push_rules_state_size_counter.inc(len(current_state_ids))
-
-            logger.debug(
-                "Looking for member changes in %r %r", state_group, current_state_ids
+            local_users = await self.store.get_local_users_in_room(
+                self.room_id, on_invalidate=self.invalidate_all_cb
```
Comment (erikjohnston) [inline synapse/push/bulk_push_rule_evaluator.py:469]: I think we can probably now remove `RoomsForUser` code and move all `get_rules` into a normal cached function, and ignore the state group stuff entirely.  

(Having a separate class with manual cache handling was only useful when we were doing things based on deltas between state groups, which we're not anymore)

File: synapse/push/bulk_push_rule_evaluator.py (hunk 1 above)
Comment (H-Shay) [reply synapse/push/bulk_push_rule_evaluator.py:316]: Any chance I could use the public version of this function: https://github.com/matrix-org/synapse/blob/13e359aec8ae8be8dc56a036ae6d9f2bc1d07385/synapse/storage/databases/main/roommember.py#L233
which seems to pull per-room profiles?

Comment (erikjohnston) [reply synapse/push/bulk_push_rule_evaluator.py:316]: So yeah, that returns the right information! The only problem with that is that it's pulling out profiles for *all* users, and ideally we're trying to reduce the amount of data we're pulling out into memory.

What we probably want is that same query, but that accepts a list of user IDs to query, using `make_in_list_sql_clause`:

https://github.com/matrix-org/synapse/blob/13e359aec8ae8be8dc56a036ae6d9f2bc1d07385/synapse/storage/databases/main/roommember.py#L426-L428

File: synapse/storage/_base.py [hunk 3]
```diff
@@ -75,6 +75,9 @@ def _invalidate_state_caches(
             self._attempt_to_invalidate_cache(
                 "get_users_in_room_with_profiles", (room_id,)
             )
+            self._attempt_to_invalidate_cache(
+                "get_number_joined_users_in_room.invalidate", (room_id,)
```
Comment (erikjohnston) [inline synapse/storage/_base.py:79]:
```suggestion
                "get_number_joined_users_in_room", (room_id,)
```

File: synapse/storage/databases/main/roommember.py [hunk 4]
```diff
@@ -444,6 +471,15 @@ def _get_rooms_for_local_user_where_membership_is_txn(
 
         return results
 
+    @cached()
+    async def get_local_users_in_room(self, room_id: str) -> List[str]:
```
Comment (erikjohnston) [inline synapse/storage/databases/main/roommember.py:475]: These need docstrings

File: synapse/push/bulk_push_rule_evaluator.py [hunk 5]
```diff
@@ -125,28 +110,42 @@ def __init__(self, hs: "HomeServer"):
         # Whether to support MSC3772 is supported.
         self._relations_match_enabled = self.hs.config.experimental.msc3772_enabled
 
+    @lru_cache()
+    async def get_rules(self, event: EventBase) -> Dict[str, List[Dict[str, dict]]]:
```
Comment (erikjohnston) [inline synapse/push/bulk_push_rule_evaluator.py:114]: The caching here isn't quite right. `lru_cache` doesn't have async support, and we'll want to cache this with a key of `room_id` rather than an `event` object. 

```suggestion
    @cached()
    async def get_rules(self, room_id: str) -> Dict[str, List[Dict[str, dict]]]:
```

File: synapse/push/bulk_push_rule_evaluator.py [hunk 6]
```diff
@@ -125,28 +110,42 @@ def __init__(self, hs: "HomeServer"):
...
+        """Given an event return the rules for all users who are
+        currently in the room.
+        """
+
+        local_users = await self.store.get_local_users_in_room(event.room_id)
```
Comment (erikjohnston) [inline synapse/push/bulk_push_rule_evaluator.py:119]: We'll want to invalidate the functions cache if `get_local_users_in_room` or `bulk_get_push_rules` cache is invalidated. We can do that using the `cache_context` machinery by a) adding a `cache_context` parameter to the function, and b) calling this with `on_invalidate=cache_context.invalidate`.

An example of this in action:

https://github.com/matrix-org/synapse/blob/d102ad67fddc650c34baa89dc7b2926d46a9aeca/synapse/storage/databases/main/appservice.py#L153-L163

File: synapse/push/bulk_push_rule_evaluator.py [hunk 7]
```diff
@@ -291,17 +284,30 @@ async def action_for_event_by_user(
...
                 continue
 
             if uid in ignorers:
                 continue
 
+            # This is a check for the case where user joins a room without being allowed to see history, and then the server
+            # receives a delayed event from before the user joined, which they should not be pushed for
+            visible = await filter_events_for_client_with_state(
+                self.store, uid, event, context
+            )
```
Comment (erikjohnston) [inline synapse/push/bulk_push_rule_evaluator.py:303]: Thinking about this some more, we probably want to pull this function call out of the loop and change it to work on a batch of users. This is a hot path and the `rules_by_user` dict could be huge, so hitting the DB one by one could cause it to take quite a lot of wall clock time.

File: synapse/storage/databases/main/roommember.py [hunk 8]
```diff
@@ -212,9 +212,9 @@ def get_users_in_room_txn(self, txn: LoggingTransaction, room_id: str) -> List[s
         txn.execute(sql, (room_id, Membership.JOIN))
         return [r[0] for r in txn]
 
-    @cached(max_entries=100000, iterable=True)
+    @cached(max_entries=100000, iterable=True, uncached_args=["user_ids"])
```
Comment (erikjohnston) [inline synapse/storage/databases/main/roommember.py:215]: `uncached_args` doesn't quite do the right thing here, as it *excluded* `user_ids` from the key used to cache the function call, so a call to get_users_in_room_with_profiles` for the same room but different sets of users will return the cached value from the first call.

Instead, we probably want to have a separate cache for this call path and use `@cachedList`, e.g. like here:

https://github.com/matrix-org/synapse/blob/d102ad67fddc650c34baa89dc7b2926d46a9aeca/synapse/storage/databases/main/keys.py#L38-L49

This *does* mean that we'd end up with two separate caches, rather than sharing one, but in this case I don't think that's a big deal as other calls to `get_users_in_room_with_profiles` are rare (and the alternative is very fiddly to write).

**Comment by erikjohnston** [comment]
Thanks Shay, this is really starting to look good! Had a quick look through what is currently there (not sure if it was ready for my review or not!), and it's mostly comments around getting the caching correct (which is always the really fiddly and brittle part). LMK if you want to jump on a call to discuss, exactly how to do caching isn't exactly obvious.

File: synapse/visibility.py [hunk 9]
```diff
@@ -160,102 +159,159 @@ def allowed(event: EventBase) -> Optional[EventBase]:
...
+    """
+    Checks to see if an event is visible to the user at the time of the event
+    Args:
+        store: databases
+        user_id: user_id to be checked
+        event: the event to be checked
+        context: EventContext for the event to be checked
+    Returns: the event if the room history is visible to the user at the time of the event, None if not
+    """
+    filter = StateFilter.from_types(
```
Comment (erikjohnston) [inline synapse/visibility.py:187]: we also need to include the check for if the event was soft failed

**Comment by erikjohnston** [comment]
Hopefully fixes https://github.com/matrix-org/synapse/issues/13178, since we'll be filtering the events for the user correctly.

File: synapse/push/bulk_push_rule_evaluator.py (hunk 5 above)
Comment (clokep) [reply synapse/push/bulk_push_rule_evaluator.py:114]: I think it is what you need, I think the following would do it:

```suggestion
    @cached(uncached_args=("event",))
    async def get_rules(self, room_id: str, event: EventBase) -> Dict[str, List[Dict[str, dict]]]:
```

It is then clearer to replace `event.room_id` internally with just `room_id`, but those should be equivalent?

How does this break the code?

Comment (H-Shay) [reply synapse/push/bulk_push_rule_evaluator.py:114]: NVMD it seems to work when I add a cache context

File: synapse/storage/databases/main/roommember.py (hunk 8 above)
Comment (H-Shay) [reply synapse/storage/databases/main/roommember.py:215]: So this I am still a little confused on-everything else should be sorted. I could either hop on a call tomorrow or maybe another explanation might help?

File: synapse/storage/databases/main/events.py [hunk 10]
```diff
@@ -1766,6 +1766,14 @@ def _store_room_members_txn(
...
                 (event.state_key,),
             )
+            txn.call_after(
+                self.store.get_local_users_in_room.invalidate,
+                (event.room_id,),
+            )
+            txn.call_after(
+                self.store.get_number_joined_users_in_room.invalidate,
+                (event.room_id,),
+            )
```
Comment (erikjohnston) [inline synapse/storage/databases/main/events.py:1776]: We need to add a cache invalidation for `get_user_in_room_with_profile` here too

File: synapse/visibility.py [hunk 11]
```diff
@@ -160,102 +159,171 @@ def allowed(event: EventBase) -> Optional[EventBase]:
...
+        return []
 
-        # if the user was a member of the room at the time of the event,
-        # they can see it.
-        if membership == Membership.JOIN:
-            return event
+    filter_list = []
+    for user_id in user_ids:
+        filter_list.append((EventTypes.Member, user_id))
+    filter_list.append((EventTypes.RoomHistoryVisibility, ""))
```
Comment (erikjohnston) [inline synapse/visibility.py:194]:
```suggestion
    filter_list = [
        (EventTypes.Member, user_id)
        for user_id in user_ids
    ]
    filter_list.append((EventTypes.RoomHistoryVisibility, ""))
```

List comprehensions can be slightly faster than a for loop.

File: synapse/visibility.py [hunk 12]
```diff
@@ -160,102 +159,171 @@ def allowed(event: EventBase) -> Optional[EventBase]:
...
+    state_map = await context.get_prev_state_ids(state_filter)
 
-        if visibility == HistoryVisibility.JOINED:
-            # we weren't a member at the time of the event, so we can't
-            # see this event.
-            return None
+    # Use events rather than event ids as content from the events are needed in _check_visibility
+    updated_state_map = {}
+    for state_key, event_id in state_map.items():
+        updated_state_map[state_key] = await store.get_event(event_id)
```
Comment (erikjohnston) [inline synapse/visibility.py:202]: We probably want to get the events in a batch using `store.get_events`

**Review by erikjohnston — COMMENTED** [review]
LGTM otherwise!

File: synapse/push/bulk_push_rule_evaluator.py [hunk 13]
```diff
@@ -125,28 +109,43 @@ def __init__(self, hs: "HomeServer"):
         # Whether to support MSC3772 is supported.
         self._relations_match_enabled = self.hs.config.experimental.msc3772_enabled
 
+    async def get_rules(
+        self, room_id: str, event: EventBase
+    ) -> Dict[str, List[Dict[str, dict]]]:
```
Comment (DMRobertson) [inline synapse/push/bulk_push_rule_evaluator.py:114]: What does the innermost dict hold?

File: synapse/push/bulk_push_rule_evaluator.py [hunk 14]
```diff
@@ -125,28 +109,43 @@ def __init__(self, hs: "HomeServer"):
...
+        """Given an event return the rules for all users who are
+        currently in the room.
```
Comment (DMRobertson) [inline synapse/push/bulk_push_rule_evaluator.py:116]: Perhaps: `for all users currently in the room who should be notified of this event`?

Has this event been persisted yet?

File: synapse/push/bulk_push_rule_evaluator.py [hunk 15]
```diff
@@ -125,28 +109,43 @@ def __init__(self, hs: "HomeServer"):
...
+        """
+
+        local_users = await self.store.get_local_users_in_room(room_id)
+
+        if event.type == EventTypes.Member and event.membership == Membership.JOIN:
+            if self.hs.is_mine_id(event.state_key):
+                local_users = list(local_users)
+                local_users.append(event.state_key)
```
Comment (DMRobertson) [inline synapse/push/bulk_push_rule_evaluator.py:124]: Signpost comment would be helpful.

If we `event.sender` is already in the room and sends another JOIN (e.g. name change), does this mean they get double-counted in `local_users`?

Comment (DMRobertson) [reply synapse/push/bulk_push_rule_evaluator.py:124]: I think the database should deduplicate though.

File: synapse/push/bulk_push_rule_evaluator.py [hunk 16]
```diff
@@ -125,28 +109,43 @@ def __init__(self, hs: "HomeServer"):
...
+
+        logger.debug("Users in room: %s", local_users)
+
+        if logger.isEnabledFor(logging.DEBUG):
+            logger.debug(
+                "Returning push rules for %r %r",
+                room_id,
+                ret_rules_by_user.keys(),
+            )
+        return ret_rules_by_user
```
Comment (DMRobertson) [inline synapse/push/bulk_push_rule_evaluator.py:136]: Either put one behind the log level guard or both, IMO?

.keys() will print `"dict_keys(['a'])"` which might not be what you tend. Perhaps `tuple(ret_rules_by_user)` on 134?

File: synapse/push/bulk_push_rule_evaluator.py [hunk 17]
```diff
@@ -125,28 +109,43 @@ def __init__(self, hs: "HomeServer"):
...
+
     async def _get_rules_for_event(
```
Comment (DMRobertson) [inline synapse/push/bulk_push_rule_evaluator.py:138]: Not sure if this needs 2 functions now (tho I like small functions :) )

File: synapse/push/bulk_push_rule_evaluator.py [hunk 18]
```diff
@@ -125,28 +109,43 @@ def __init__(self, hs: "HomeServer"):
...
-            hs=self.hs,
-            room_id=room_id,
-            rules_for_room_cache=self._get_rules_for_room.cache,
-            room_push_rule_cache_metrics=self.room_push_rule_cache_metrics,
-            linearizer=self._rules_linearizer,
-            cached_data=rules_for_room_data,
-        )
-
-        rules_by_user = await rules_for_room.get_rules(event, context)
+        rules_by_user = await self.get_rules(event.room_id, event)
```
Comment (DMRobertson) [inline synapse/push/bulk_push_rule_evaluator.py:148]: May as well move the logic below above, so that we make one DB larger call rather than two mediumer ones.

File: synapse/push/bulk_push_rule_evaluator.py [hunk 19]
```diff
@@ -278,30 +271,35 @@ async def action_for_event_by_user(
...
 
-        # If the event is not a state event check if any users ignore the sender.
-        if not event.is_state():
-            ignorers = await self.store.ignored_by(event.sender)
-        else:
-            ignorers = frozenset()
+        users = list(rules_by_user.keys())
+        profiles = await self.store.get_subset_users_in_room_with_profiles(
+            event.room_id, users
+        )
```
Comment (DMRobertson) [inline synapse/push/bulk_push_rule_evaluator.py:284]: I thought the `list` was weird (but fine). It will allocate though.

If you're a microoptimising speed freak you could pass the dict directly. Or pass `dict.keys()` directly. Change the store function to accept a Collection[str] (which is what `make_in_list_sql_clause` actually wants).

File: synapse/push/bulk_push_rule_evaluator.py [hunk 20]
```diff
@@ -125,28 +109,43 @@ def __init__(self, hs: "HomeServer"):
...
         self._relations_match_enabled = self.hs.config.experimental.msc3772_enabled
 
+    async def get_rules(
+        self, room_id: str, event: EventBase
+    ) -> Dict[str, List[Dict[str, dict]]]:
+        """Given an event return the rules for all users who are
+        currently in the room.
+        """
+
+        local_users = await self.store.get_local_users_in_room(room_id)
```
Comment (DMRobertson) [inline synapse/push/bulk_push_rule_evaluator.py:119]: Insight from E:

"Find out who is in the room NOW, and then what they can see" is cheaper than "find out who was in the room when the event was received" because

- who is in the room now is immediately available and doesn't require fetching state
- there are probably relatively few local users compared to loads of remote users in a large room.

Might be nice to have in a docstring or comment somewhere.

**Review by DMRobertson — COMMENTED** [review]
TL;DR: nothing looks wrong; it looks simpler; it seems sane. Nitpicks and general reviewer arsey remains.

**Review by DMRobertson — APPROVED** [review]
Sytest failure looks to be https://github.com/matrix-org/sytest/issues/1125.

LGTM otherwise; I'm happy if CI is happy.

**Review by Shudogg — COMMENTED** [review]
Msg
